**Задание для проекта**
- В качестве датасета используем данные онлайн-кинотеатра. Имеется информация о пользователях, фильмах, а так же оценках, которые пользователи поставили тому или иному фильму.
- В рамках проекта хотим провести исследование текущей ситуации и решить бизнес-кейс с рекомендациями фильмов пользователям (очевидно, что не все пользователи смотрели все фильмы и нам нужно каким-то образом рекомендовать пользователю, что ему посмотреть следующим)

- Датасет доступен тут - https://grouplens.org/datasets/movielens/100k/
- Описание тут - http://files.grouplens.org/datasets/movielens/ml-100k-README.txt


1. Загружаем в колаб файлы по оценкам (ratings) и фильмам (movies) и создаем на их основе pandas-датафреймы

In [ ]:
import pandas as pd

In [ ]:
# Загружаем файл по рейтингам
df = pd.read_csv('rating.csv')
df.head()

ID user  ID film  rating  timestsmp                 date
0      196      242       3  881250949  04.12.1997 15:55:49
1      186      302       3  891717742  04.04.1998 19:22:22
2       22      377       1  878887116   07.11.1997 7:18:36
3      244       51       2  880606923   27.11.1997 5:02:03
4      166      346       1  886397596   02.02.1998 5:33:16

In [ ]:
len(df)

100000

In [ ]:
# Загружаем файл по фильмам
df_f = pd.read_csv('movies.csv')
df_f.head()

ID film              Title Release date  \
0        1   Toy Story (1995)   01.01.1995   
1        2   GoldenEye (1995)   01.01.1995   
2        3  Four Rooms (1995)   01.01.1995   
3        4  Get Shorty (1995)   01.01.1995   
4        5     Copycat (1995)   01.01.1995   

                                                 URL  Неизвестен  Боевик  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...           0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...           0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...           0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...           0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)           0       0   

   Приключения  Анимация  Детский  Комедия  ...  Фентези  Нуар  Ужасы  Мюзикл  \
0            0         1        1        1  ...        0     0      0       0   
1            1         0        0        0  ...        0     0      0       0   
2            0         0        0        0  ...        0     0      0       0   
3            0         0        0        1  ...        0     0      0       0   
4            0         0        0        0  ...        0     0      0       0   

   Мистика  Романтика  Фантастика  Триллер  Война  Вестерн  
0        0          0           0        0      0        0  
1        0          0           0        1      0        0  
2        0          0           0        1      0        0  
3        0          0           0        0      0        0  
4        0          0           0        1      0        0  

[5 rows x 23 columns]

In [ ]:
len(df_f)

1682

2. Средствами Pandas, используя dataframe ratings, находим id пользователя, поставившего больше всего оценок



In [ ]:
df_1 = df['ID user'].value_counts().reset_index().head(1)
df_1


index  ID user
0    405      737

Пользователь с ID = 405 поставил больше всего оценок - 737

3. Оставляем в датафрейме ratings только те фильмы, который оценил данный пользователь

In [ ]:
df_2 = df[(df['ID user'] == 405)]
df_2

ID user  ID film  rating  timestsmp                date
12276      405       56       4  885544911  23.01.1998 8:41:51
12383      405      592       1  885548670  23.01.1998 9:44:30
12430      405     1582       1  885548670  23.01.1998 9:44:30
12449      405      171       1  885549544  23.01.1998 9:59:04
12460      405      580       1  885547447  23.01.1998 9:24:07
...        ...      ...     ...        ...                 ...
98956      405      375       1  885546835  23.01.1998 9:13:55
98978      405      445       4  885548435  23.01.1998 9:40:35
99148      405     1246       1  885547735  23.01.1998 9:28:55
99465      405      196       1  885546112  23.01.1998 9:01:52
99756      405      958       1  885549590  23.01.1998 9:59:50

[737 rows x 5 columns]

4. Добавляем к датафрейму из задания 3 столбцы:
- По жанрам. Каждый столбец - это жанр. Единицу записываем, если фильм принадлежит данному жанру и 0 - если нет.
- Столбцы с общим количеством оценок от всех пользователей на фильм и суммарной оценкой от всех пользователей


In [ ]:
# Объединяем таблицы рейтингов и фильмов левым соединением по ключу ID film
df_3 = df_2.merge(df_f, how='left', on='ID film')
df_3.head()

ID user  ID film  rating  timestsmp                date  \
0      405       56       4  885544911  23.01.1998 8:41:51   
1      405      592       1  885548670  23.01.1998 9:44:30   
2      405     1582       1  885548670  23.01.1998 9:44:30   
3      405      171       1  885549544  23.01.1998 9:59:04   
4      405      580       1  885547447  23.01.1998 9:24:07   

                                               Title Release date  \
0                                Pulp Fiction (1994)   01.01.1994   
1                                  True Crime (1995)   01.01.1995   
2                                       T-Men (1947)   01.01.1947   
3                                Delicatessen (1991)   01.01.1991   
4  Englishman Who Went Up a Hill, But Came Down a...   01.01.1995   

                                                 URL  Неизвестен  Боевик  ...  \
0  http://us.imdb.com/M/title-exact?Pulp%20Fictio...           0       0  ...   
1  http://us.imdb.com/M/title-exact?True%20Crime%...           0       0  ...   
2    http://us.imdb.com/M/title-exact?T-Men%20(1947)           0       0  ...   
3  http://us.imdb.com/M/title-exact?Delicatessen%...           0       0  ...   
4  http://us.imdb.com/M/title-exact?Englishman%20...           0       0  ...   

   Фентези  Нуар  Ужасы  Мюзикл  Мистика  Романтика  Фантастика  Триллер  \
0        0     0      0       0        0          0           0        0   
1        0     0      0       0        1          0           0        1   
2        0     1      0       0        0          0           0        0   
3        0     0      0       0        0          0           1        0   
4        0     0      0       0        0          1           0        0   

   Война  Вестерн  
0      0        0  
1      0        0  
2      0        0  
3      0        0  
4      0        0  

[5 rows x 27 columns]

In [ ]:
# Общее количество оценок и суммарная оценка от всех пользователей на фильм
df_4 = df.groupby('ID film').agg(['count', 'sum'])['rating'].reset_index()
df_4.head()

ID film  count   sum
0        1    452  1753
1        2    131   420
2        3     90   273
3        4    209   742
4        5     86   284

In [ ]:
# Cоединяем снова полученные данные
df_5 = df_3.merge(df_4, how='left', on='ID film')
df_5.head()


ID user  ID film  rating  timestsmp                date  \
0      405       56       4  885544911  23.01.1998 8:41:51   
1      405      592       1  885548670  23.01.1998 9:44:30   
2      405     1582       1  885548670  23.01.1998 9:44:30   
3      405      171       1  885549544  23.01.1998 9:59:04   
4      405      580       1  885547447  23.01.1998 9:24:07   

                                               Title Release date  \
0                                Pulp Fiction (1994)   01.01.1994   
1                                  True Crime (1995)   01.01.1995   
2                                       T-Men (1947)   01.01.1947   
3                                Delicatessen (1991)   01.01.1991   
4  Englishman Who Went Up a Hill, But Came Down a...   01.01.1995   

                                                 URL  Неизвестен  Боевик  ...  \
0  http://us.imdb.com/M/title-exact?Pulp%20Fictio...           0       0  ...   
1  http://us.imdb.com/M/title-exact?True%20Crime%...           0       0  ...   
2    http://us.imdb.com/M/title-exact?T-Men%20(1947)           0       0  ...   
3  http://us.imdb.com/M/title-exact?Delicatessen%...           0       0  ...   
4  http://us.imdb.com/M/title-exact?Englishman%20...           0       0  ...   

   Ужасы  Мюзикл  Мистика  Романтика  Фантастика  Триллер  Война  Вестерн  \
0      0       0        0          0           0        0      0        0   
1      0       0        1          0           0        1      0        0   
2      0       0        0          0           0        0      0        0   
3      0       0        0          0           1        0      0        0   
4      0       0        0          1           0        0      0        0   

   count   sum  
0    394  1600  
1      9    30  
2      1     1  
3     65   252  
4     32   108  

[5 rows x 29 columns]

Разбиваем данные на множество для обучения и для проверки

Берем модель линейной регрессии и обучаем ее



In [ ]:
# Выводим названия столбцов датафрейма
df_5.columns

Index(['ID user', 'ID film', 'rating', 'timestsmp', 'date', 'Title',
       'Release date', 'URL', 'Неизвестен', 'Боевик', 'Приключения',
       'Анимация', 'Детский', 'Комедия', 'Криминал', 'Документальный', 'Драма',
       'Фентези', 'Нуар', 'Ужасы', 'Мюзикл', 'Мистика', 'Романтика',
       'Фантастика', 'Триллер', 'Война', 'Вестерн', 'count', 'sum'],
      dtype='object')

In [ ]:
df_lda = df_5[['Неизвестен', 'Боевик', 'Приключения', 'Анимация', 'Детский', 'Комедия', 'Криминал', 'Документальный', 'Драма',
       'Фентези', 'Нуар', 'Ужасы', 'Мюзикл', 'Мистика', 'Романтика',
       'Фантастика', 'Триллер', 'Война', 'Вестерн', 'count', 'sum']]

In [ ]:
df_class = df_5['rating']

In [ ]:
# Импортируем библиотеку
import sklearn

In [ ]:
# разделяем выборку на тренировочную и тестовую
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_lda, df_class, random_state=40)

In [ ]:
# импортируем LDA-функционал
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [ ]:
lda = LinearDiscriminantAnalysis()

In [ ]:
# обучаем данные
lda.fit(X_train, y_train)

LinearDiscriminantAnalysis()

In [ ]:
# делаем прогноз на тестовой выборке
lda.predict(X_test)

array([1, 1, 1, 1, 1, 1, 1, 1, 4, 5, 1, 1, 1, 1, 5, 1, 5, 1, 1, 1, 1, 1,
       5, 1, 4, 1, 2, 4, 1, 1, 1, 4, 5, 1, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 5, 1, 1, 5, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1,
       5, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 2, 1, 1,
       1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 5, 1, 4,
       5, 1, 1, 4, 5, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1,
       1, 1, 1, 5, 1, 1, 1, 1, 2, 1, 1, 1, 5, 1, 1, 5, 1, 1, 5, 1, 1, 1,
       5, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 5, 1, 1, 1, 1, 1, 1, 4, 1, 1,
       1, 1, 4, 1, 1, 1, 1, 1, 1])

In [ ]:
# смотрим разницу факта и прогноза
result = pd.DataFrame([y_test, lda.predict(X_test)]).T

In [ ]:
result

rating  Unnamed 0
385     1.0        NaN
464     4.0        NaN
335     1.0        NaN
331     5.0        NaN
333     3.0        NaN
..      ...        ...
392     1.0        NaN
733     4.0        NaN
690     3.0        NaN
319     1.0        NaN
26      1.0        2.0

[185 rows x 2 columns]

In [ ]:
# Оценим качество модели метрикой точности
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(y_test, lda.predict(X_test))

0.6108108108108108

С учетом имеющихся данных, модель, предсказывающая оценку фильма пользователем имеет точность 61% на тестовой выборке.

**Загружаем данные в Spark**

In [ ]:
!apt-get update

Get:1 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]
Hit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]
Get:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [109 kB]
Hit:7 https://ppa.launchpadcontent.net/c2d4u.team/c2d4u4.0+/ubuntu jammy InRelease
Get:8 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [1,342 kB]
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Get:10 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,266 kB]
Hit:11 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Fetched 2,950 kB in 2s (1,715 kB/s)
Reading package li

In [ ]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null

In [ ]:
!wget -q https://downloads.apache.org/spark/spark-3.5.0/spark-3.5.0-bin-hadoop3.tgz

In [ ]:
!tar -xvf spark-3.5.0-bin-hadoop3.tgz

spark-3.5.0-bin-hadoop3/
spark-3.5.0-bin-hadoop3/kubernetes/
spark-3.5.0-bin-hadoop3/kubernetes/tests/
spark-3.5.0-bin-hadoop3/kubernetes/tests/pyfiles.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/decommissioning.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/autoscale.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/python_executable_check.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/worker_memory_check.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/py_container_checks.py
spark-3.5.0-bin-hadoop3/kubernetes/tests/decommissioning_cleanup.py
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/decom.sh
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/bindings/
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/bindings/R/
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/bindings/R/Dockerfile
spark-3.5.0-bin-hadoop3/kubernetes/dockerfiles/spark/bindings/python/
spark-3.5.0-bin-hadoop3/kubern

In [ ]:
!pip install -q findspark

In [ ]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.5.0-bin-hadoop3"

In [ ]:
import findspark
findspark.init()
from pyspark.sql import SparkSession

spark = SparkSession.builder.master("local[*]").getOrCreate()

In [ ]:
# Загружаем данные по рейтингам в Spark
df_rating = spark.read.csv('rating.csv', header=True)

In [ ]:
df_rating.show(5)

+-------+-------+------+---------+-------------------+
|ID user|ID film|rating|timestsmp|               date|
+-------+-------+------+---------+-------------------+
|    196|    242|     3|881250949|04.12.1997 15:55:49|
|    186|    302|     3|891717742|04.04.1998 19:22:22|
|     22|    377|     1|878887116| 07.11.1997 7:18:36|
|    244|     51|     2|880606923| 27.11.1997 5:02:03|
|    166|    346|     1|886397596| 02.02.1998 5:33:16|
+-------+-------+------+---------+-------------------+
only showing top 5 rows



In [ ]:
# Загружаем данные по фильмам в Spark
df_movies = spark.read.csv('movies.csv', header=True)

In [ ]:
df_movies.show(5)

+-------+-----------------+------------+--------------------+----------+------+-----------+--------+-------+-------+--------+--------------+-----+-------+----+-----+------+-------+---------+----------+-------+-----+-------+
|ID film|            Title|Release date|                 URL|Неизвестен|Боевик|Приключения|Анимация|Детский|Комедия|Криминал|Документальный|Драма|Фентези|Нуар|Ужасы|Мюзикл|Мистика|Романтика|Фантастика|Триллер|Война|Вестерн|
+-------+-----------------+------------+--------------------+----------+------+-----------+--------+-------+-------+--------+--------------+-----+-------+----+-----+------+-------+---------+----------+-------+-----+-------+
|      1| Toy Story (1995)|  01.01.1995|http://us.imdb.co...|         0|     0|          0|       1|      1|      1|       0|             0|    0|      0|   0|    0|     0|      0|        0|         0|      0|    0|      0|
|      2| GoldenEye (1995)|  01.01.1995|http://us.imdb.co...|         0|     1|          1|       0|    

In [ ]:
# Изменим тип данных на float
df1_rating = df_rating.withColumn('ID user', df_rating['ID user'].cast('float'))

In [ ]:
df2_rating = df1_rating.withColumn('ID film', df_rating['ID film'].cast('float'))

In [ ]:
df3_rating = df2_rating.withColumn('rating', df_rating['rating'].cast('float'))

In [ ]:
df1_movies = df_movies.withColumn('Неизвестен', df_movies['Неизвестен'].cast('float'))

In [ ]:
df2_movies = df1_movies.withColumn('Боевик', df_movies['Боевик'].cast('float'))

In [ ]:
df3_movies = df2_movies.withColumn('Приключения', df_movies['Приключения'].cast('float'))

In [ ]:
df4_movies = df3_movies.withColumn('Анимация', df_movies['Анимация'].cast('float'))

In [ ]:
df5_movies = df4_movies.withColumn('Детский', df_movies['Детский'].cast('float'))

In [ ]:
df6_movies = df5_movies.withColumn('Комедия', df_movies['Комедия'].cast('float'))

In [ ]:
df7_movies = df6_movies.withColumn('Криминал', df_movies['Криминал'].cast('float'))

In [ ]:
df8_movies = df7_movies.withColumn('Документальный', df_movies['Документальный'].cast('float'))

In [ ]:
df9_movies = df8_movies.withColumn('Драма', df_movies['Драма'].cast('float'))

In [ ]:
df10_movies = df9_movies.withColumn('Фентези', df_movies['Фентези'].cast('float'))

In [ ]:
df11_movies = df10_movies.withColumn('Нуар', df_movies['Нуар'].cast('float'))

In [ ]:
df12_movies = df11_movies.withColumn('Ужасы', df_movies['Ужасы'].cast('float'))

In [ ]:
df13_movies = df12_movies.withColumn('Мюзикл', df_movies['Мюзикл'].cast('float'))

In [ ]:
df14_movies = df13_movies.withColumn('Мистика', df_movies['Мистика'].cast('float'))

In [ ]:
df15_movies = df14_movies.withColumn('Романтика', df_movies['Романтика'].cast('float'))

In [ ]:
df16_movies = df15_movies.withColumn('Фантастика', df_movies['Фантастика'].cast('float'))

In [ ]:
df17_movies = df16_movies.withColumn('Триллер', df_movies['Триллер'].cast('float'))

In [ ]:
df18_movies = df17_movies.withColumn('Война', df_movies['Война'].cast('float'))

In [ ]:
df19_movies = df18_movies.withColumn('Вестерн', df_movies['Вестерн'].cast('float'))

In [ ]:
# Средняя оценка для каждого фильма
df_film = df3_rating.groupby('ID film').mean('rating')
df_film.show(5)



+-------+-----------------+
|ID film|      avg(rating)|
+-------+-----------------+
| 1033.0|          2.78125|
|  714.0|3.230769230769231|
| 1575.0|              1.0|
| 1081.0|             2.75|
| 1393.0|2.111111111111111|
+-------+-----------------+
only showing top 5 rows



In [ ]:
# Соединяем данные по фильмам и оценкам

df_full = df3_rating.join(df19_movies, 'ID film', how='left')
df_full.show(5)

+-------+-------+------+---------+-------------------+--------------------+------------+--------------------+----------+------+-----------+--------+-------+-------+--------+--------------+-----+-------+----+-----+------+-------+---------+----------+-------+-----+-------+
|ID film|ID user|rating|timestsmp|               date|               Title|Release date|                 URL|Неизвестен|Боевик|Приключения|Анимация|Детский|Комедия|Криминал|Документальный|Драма|Фентези|Нуар|Ужасы|Мюзикл|Мистика|Романтика|Фантастика|Триллер|Война|Вестерн|
+-------+-------+------+---------+-------------------+--------------------+------------+--------------------+----------+------+-----------+--------+-------+-------+--------+--------------+-----+-------+----+-----+------+-------+---------+----------+-------+-----+-------+
|  242.0|  196.0|   3.0|881250949|04.12.1997 15:55:49|        Kolya (1996)|  24.01.1997|http://us.imdb.co...|       0.0|   0.0|        0.0|     0.0|    0.0|    1.0|     0.0|           

In [ ]:
# Средняя оценка жанра Неизвестен
df_genre1 = df_full.where((df_full['Неизвестен']==1))
df_genre1.groupby('Неизвестен').mean('rating').show()

+----------+-----------+
|Неизвестен|avg(rating)|
+----------+-----------+
|       1.0|        3.2|
+----------+-----------+



In [ ]:
# Средняя оценка жанра Боевик
df_genre2 = df_full.where((df_full['Боевик']==1))
df_genre2.groupby('Боевик').mean('rating').show()

+------+-----------------+
|Боевик|      avg(rating)|
+------+-----------------+
|   1.0|3.480245417953027|
+------+-----------------+



In [ ]:
# Средняя оценка жанра Приключения
df_genre3 = df_full.where((df_full['Приключения']==1))
df_genre3.groupby('Приключения').mean('rating').show()

+-----------+-----------------+
|Приключения|      avg(rating)|
+-----------+-----------------+
|        1.0|3.503526503308369|
+-----------+-----------------+



In [ ]:
# Средняя оценка жанра Анимация
df_genre4 = df_full.where((df_full['Анимация']==1))
df_genre4.groupby('Анимация').mean('rating').show()

+--------+------------------+
|Анимация|       avg(rating)|
+--------+------------------+
|     1.0|3.5766990291262135|
+--------+------------------+



In [ ]:
# Средняя оценка жанра Детский
df_genre5 = df_full.where((df_full['Детский']==1))
df_genre5.groupby('Детский').mean('rating').show()

+-------+------------------+
|Детский|       avg(rating)|
+-------+------------------+
|    1.0|3.3532442216652742|
+-------+------------------+



In [ ]:
# Средняя оценка жанра Комедия
df_genre6 = df_full.where((df_full['Комедия']==1))
df_genre6.groupby('Комедия').mean('rating').show()

+-------+------------------+
|Комедия|       avg(rating)|
+-------+------------------+
|    1.0|3.3940734781442745|
+-------+------------------+



In [ ]:
# Средняя оценка жанра Криминал
df_genre7 = df_full.where((df_full['Криминал']==1))
df_genre7.groupby('Криминал').mean('rating').show()

+--------+------------------+
|Криминал|       avg(rating)|
+--------+------------------+
|     1.0|3.6322780881440098|
+--------+------------------+



In [ ]:
# Средняя оценка жанра Документальный
df_genre8 = df_full.where((df_full['Документальный']==1))
df_genre8.groupby('Документальный').mean('rating').show()

+--------------+------------------+
|Документальный|       avg(rating)|
+--------------+------------------+
|           1.0|3.6728232189973613|
+--------------+------------------+



In [ ]:
# Средняя оценка жанра Драма
df_genre9 = df_full.where((df_full['Драма']==1))
df_genre9.groupby('Драма').mean('rating').show()

+-----+------------------+
|Драма|       avg(rating)|
+-----+------------------+
|  1.0|3.6873793708484772|
+-----+------------------+



In [ ]:
# Средняя оценка жанра Фентези
df_genre10 = df_full.where((df_full['Фентези']==1))
df_genre10.groupby('Фентези').mean('rating').show()

+-------+------------------+
|Фентези|       avg(rating)|
+-------+------------------+
|    1.0|3.2152366863905324|
+-------+------------------+



In [ ]:
# Средняя оценка жанра Нуар
df_genre11 = df_full.where((df_full['Нуар']==1))
df_genre11.groupby('Нуар').mean('rating').show()

+----+------------------+
|Нуар|       avg(rating)|
+----+------------------+
| 1.0|3.9215233698788228|
+----+------------------+



In [ ]:
# Средняя оценка жанра Ужасы
df_genre12 = df_full.where((df_full['Ужасы']==1))
df_genre12.groupby('Ужасы').mean('rating').show()

+-----+------------------+
|Ужасы|       avg(rating)|
+-----+------------------+
|  1.0|3.2903893172841827|
+-----+------------------+



In [ ]:
# Средняя оценка жанра Документальный
df_genre13 = df_full.where((df_full['Мюзикл']==1))
df_genre13.groupby('Мюзикл').mean('rating').show()

+------+-----------------+
|Мюзикл|      avg(rating)|
+------+-----------------+
|   1.0|3.521396851029471|
+------+-----------------+



In [ ]:
# Средняя оценка жанра Мистика
df_genre14 = df_full.where((df_full['Мистика']==1))
df_genre14.groupby('Мистика').mean('rating').show()

+-------+----------------+
|Мистика|     avg(rating)|
+-------+----------------+
|    1.0|3.63813155386082|
+-------+----------------+



In [ ]:
# Средняя оценка жанра Романтика
df_genre15 = df_full.where((df_full['Романтика']==1))
df_genre15.groupby('Романтика').mean('rating').show()

+---------+-----------------+
|Романтика|      avg(rating)|
+---------+-----------------+
|      1.0|3.621704948358255|
+---------+-----------------+



In [ ]:
# Средняя оценка жанра Фантастика
df_genre16 = df_full.where((df_full['Фантастика']==1))
df_genre16.groupby('Фантастика').mean('rating').show()

+----------+------------------+
|Фантастика|       avg(rating)|
+----------+------------------+
|       1.0|3.5607227022780834|
+----------+------------------+



In [ ]:
# Средняя оценка жанра Триллер
df_genre17 = df_full.where((df_full['Триллер']==1))
df_genre17.groupby('Триллер').mean('rating').show()

+-------+------------------+
|Триллер|       avg(rating)|
+-------+------------------+
|    1.0|3.5090069495245064|
+-------+------------------+



In [ ]:
# Средняя оценка жанра Война
df_genre18 = df_full.where((df_full['Война']==1))
df_genre18.groupby('Война').mean('rating').show()

+-----+-----------------+
|Война|      avg(rating)|
+-----+-----------------+
|  1.0|3.815811874866993|
+-----+-----------------+



In [ ]:
# Средняя оценка жанра Вестерн
df_genre19 = df_full.where((df_full['Вестерн']==1))
df_genre19.groupby('Вестерн').mean('rating').show()



+-------+------------------+
|Вестерн|       avg(rating)|
+-------+------------------+
|    1.0|3.6132686084142396|
+-------+------------------+



Получаем 2 датафрейма с 5-ю самыми популярными и самыми непопулярными фильмами по количеству оценок

In [ ]:
# Топ 5 популярных фильмов
df_top = df_full.groupby('ID film', 'Title').sum('rating').sort(['sum(rating)'], ascending=False)
df_top.show(5)

+-------+--------------------+-----------+
|ID film|               Title|sum(rating)|
+-------+--------------------+-----------+
|   50.0|    Star Wars (1977)|     2541.0|
|  100.0|        Fargo (1996)|     2111.0|
|  181.0|Return of the Jed...|     2032.0|
|  258.0|      Contact (1997)|     1936.0|
|  174.0|Raiders of the Lo...|     1786.0|
+-------+--------------------+-----------+
only showing top 5 rows



In [ ]:
# Топ 5 непопулярных фильмов
df_top = df_full.groupby('ID film', 'Title').sum('rating').sort(['sum(rating)'])
df_top.show(5)

+-------+--------------------+-----------+
|ID film|               Title|sum(rating)|
+-------+--------------------+-----------+
| 1561.0|Tigrero: A Film T...|        1.0|
| 1343.0|   Lotto Land (1995)|        1.0|
| 1626.0|Nobody Loves Me (...|        1.0|
| 1557.0|  Yankee Zulu (1994)|        1.0|
| 1579.0|  Baton Rouge (1988)|        1.0|
+-------+--------------------+-----------+
only showing top 5 rows

